src/adapter/etl-adapter-google-sheet/src/Flow/ETL/Adapter/GoogleSheet/GoogleSheetLoader.php

```diff
@@ -0,0 +1,179 @@
+    }
+
+    public function load(Rows $rows, FlowContext $context) : void
+    {
+        if (!$rows->count()) {
+            return;
+        }
+
+        $columnsCount = $rows->first()->entries()->count();
+        $currentRow = $this->calculateCurrentRow($rows);
+        $sheetRange = $this->calculateSheetAddress($columnsCount, $currentRow);
+
+        $this->increaseSheetSize($currentRow, $columnsCount);
+
+        $normalizer = new RowsNormalizer(
+            new EntryNormalizer($context->config->caster(), $this->dateTimeFormat, $this->dateFormat)
+        );
+
+        $this->service->spreadsheets_values->update(
+            $this->spreadsheetId,
+            $sheetRange->toString(),
+            new ValueRange(
+                [
+                    'values' => $this->values($rows, $normalizer),
+                    'majorDimension' => 'ROWS',
+                ]
+            ),
+            $this->inputOption->toArray()
+        );
+
+        $this->loadedRows = $currentRow;
+    }
+
+    public function withDateFormat(string $dateFormat) : self
+    {
+        $this->dateFormat = $dateFormat;
+
+        return $this;
+    }
+
+    public function withDateTimeFormat(string $dateTimeFormat) : self
+    {
+        $this->dateTimeFormat = $dateTimeFormat;
+
+        return $this;
+    }
+
+    public function withHeader(bool $withHeader) : self
+    {
+        $this->withHeader = $withHeader;
+
+        return $this;
+    }
+
+    public function withInputOption(ValueInputOption $inputOption) : self
+    {
+        $this->inputOption = $inputOption;
+
+        return $this;
+    }
+
+    private function calculateCurrentRow(Rows $rows) : int
+    {
+        return $rows->count() + (0 === $this->loadedRows ? ($this->withHeader ? 1 : 0) : $this->loadedRows);
+    }
+
+    private function calculateSheetAddress(int $columnsCount, int $currentRow) : SheetAddress
+    {
+        return SheetAddress::calculate(
+            $columnsCount,
+            $currentRow,
+            new SheetAddress(
+                startCell: 0 === $this->loadedRows ? 'A1' : 'A' . $this->loadedRows + 1,
+                sheetName: $this->sheetName
+            )
+        );
+    }
+
+    private function increaseSheetSize(int $currentRow, int $columnsCount) : void
+    {
+        // Check if we need to increase sheet size from default 5k rows
+        if (!$this->isDividable($currentRow) && !$this->isDividable($currentRow - 1)) {
+            return;
+        }
+
+        $this->service->spreadsheets->batchUpdate(
+            $this->spreadsheetId,
+            new BatchUpdateSpreadsheetRequest([
+                'requests' => [
+                    [
+                        'updateSheetProperties' => [
+                            'properties' => [
+                                'sheetId' => $this->sheetId(),
+                                'gridProperties' => [
+                                    'rowCount' => $currentRow + 5_000,
```

Comment:
shouldn't this just add count of loaded Rows? Otherwise it will just keep allocating 5k rows even when it's not necessary